archive/extract.go

```diff
@@ -15,11 +15,12 @@ type PathMode struct {
 	Mode os.FileMode
 }
 
-// Extract reads all entries from TarReader and extracts them to the filesystem
+// Extract reads all entries from TarReader and extracts them to the filesystem.
+// Note that this function is not thread-safe on Unix as it mutates the global umask.
 func Extract(tr TarReader) error {
```

Comment:
```suggestion
var umask int

func init() {
	umask = setUmask(0)
	setUmask(umask)
}

// Extract reads all entries from TarReader and extracts them to the filesystem
func Extract(tr TarReader) error {
```

Unfortunately, I'm still able to reproduce @ryanmoran's case in this PR. What do you think of the following?

Reply:
I'm a bit wary to set the umask for any program that might import this package (for the entirety of the program's execution). Are you using the linked [reproduction program](https://gist.github.com/ryanmoran/44b9e8fe1bbd6b59a86a2626ec1275a7#file-main-go) to verify? If so I think it makes sense that you're still seeing `umask after: 0` in the output. [These lines](https://github.com/buildpacks/lifecycle/pull/722/files#diff-98f352cc6ef347aa505fe6d5f3e01c3a47ee0b5d1a91d1e918343284c893b51eR44-R45) are a bit like [this line](https://gist.github.com/ryanmoran/44b9e8fe1bbd6b59a86a2626ec1275a7#file-main-go-L24) in the reproduction program, setting the umask back to whatever it was before `archive.Extract` was called.

Reply:
I think the problem is that we're trying to solve is that we want a `fetchUmask()`, so we can use it for the Extract() operation.

But it looks like the only way to _fetch_ is to `setUmask()` to get the old value. And then set it again.

The problem with our `defer` approach is that the Extract() can take a long time. Thus leaving the "gap" (_when the throwaway value is set_) to be too long. Which wouldn't be good for both race conditions, and in-between syscalls to `open()` or `mkdir()`